ngramoffset.go

```diff
@@ -112,3 +122,211 @@ func (a *arrayNgramOffset) DumpMap() map[ngram]simpleSection {
+// a 40b packing with 19-bit lengths, but the code would be significantly uglier so it doesn't
+// seem worth it.
+const ngramAsciiMaxSectionLength = (1 << 11) - 1
+
+func ngramAsciiToPacked(ng ngram) uint32 {
+	return uint32(ng&127) | uint32((ng>>(21-7))&(127<<7)) | uint32((ng>>(42-14))&(127<<14))
+}
+
+func ngramAsciiPackedToNgram(ng uint32) ngram {
+	return ngram(ng&127) | ngram(ng&(127<<7))<<(21-7) | ngram(ng&(127<<14))<<(42-14)
+}
+
+// asciiNgramOffset stores ascii trigrams packed together with short lengths,
+// using offsets for a chunk of entries to limit the number of lengths that must
+// be summed to compute a section's offset.
+type asciiNgramOffset struct {
+	entries      []uint32 // (chara << 25 | charb << 18 | charc << 11 | length)
+	chunkOffsets []uint32 // offset for entries[i*asciiNgramOffsetChunkLength]
+}
+
+// asciiNgramOffsetChunkLength specifies how many entries share one initial offset.
+// It must be a power of 2, and was chosen empirically by measuring RAM usage:
+// 8: 4132MB, 16: 4047MB, 32: 4006MB, 64: 3992MB, 128: 3990MB
+const asciiNgramOffsetChunkLength = 32
+
+func makeAsciiNgramOffset(ngrams []uint32, offsets []uint32) *asciiNgramOffset {
+	// fmt.Println("makeAsciiNgramOffset", ngrams, offsets)
+	ao := &asciiNgramOffset{
+		entries:      make([]uint32, 0, len(ngrams)),
+		chunkOffsets: make([]uint32, 0, len(ngrams)/asciiNgramOffsetChunkLength),
+	}
+
+	for i, ng := range ngrams {
+		if len(ao.entries)%asciiNgramOffsetChunkLength == 0 {
+			ao.chunkOffsets = append(ao.chunkOffsets, offsets[i])
+		}
+		length := offsets[i+1] - offsets[i]
+		for length > 0 {
+			if length >= ngramAsciiMaxSectionLength {
+				ao.entries = append(ao.entries, ng<<11|ngramAsciiMaxSectionLength)
+				length -= ngramAsciiMaxSectionLength
+			} else {
+				ao.entries = append(ao.entries, ng<<11|length)
+				length = 0
+			}
+			if len(ao.entries)%asciiNgramOffsetChunkLength == 0 {
+				break
+			}
+		}
+	}
+
+	ao.entries = shrinkUint32Slice(ao.entries)
+	ao.chunkOffsets = shrinkUint32Slice(ao.chunkOffsets)
+
+	return ao
+}
+
+// Get returns a simpleSection with sz=0 if no entry, or sz=ngramAsciiMaxSectionLength
+// if the length of the ngram is too large for this type and it should cascade to the next entry.
+func (a *asciiNgramOffset) Get(gram ngram) simpleSection {
+	if gram&ngramAsciiMask != gram {
+		return simpleSection{}
+	}
+	g := ngramAsciiToPacked(gram) << 11
+
+	idx := sort.Search(len(a.entries), func(i int) bool {
+		return a.entries[i] >= g
+	})
+
+	if idx == len(a.entries) || a.entries[idx]>>11 != g>>11 {
+		return simpleSection{}
+	}
+
+	length := a.entries[idx] & ngramAsciiMaxSectionLength
+	if length == ngramAsciiMaxSectionLength {
+		// this ascii ngram's section length is too large to be represented;
+		// repeate the Get() on the unicode map to get the correct result.
+		return simpleSection{
+			off: 0,
+			sz:  ngramAsciiMaxSectionLength,
```

Comment:
idea: don't special case `ngramAsciiMaxSectionLength` in the combined get. Rather just return 0 here and pretend we don't have the value. Simplifies invariants that combined needs to care about.

Reply:
That would force two lookups for every ascii ngram that isn't present instead of just one.